source/main/physics/BeamFactory.cpp

```diff
@@ -481,83 +481,105 @@ Beam *BeamFactory::getBeam(int source_id, int stream_id)
-			if (it->intersects(m_trucks[a]->boundingBox))
+			if (intersectionAABB(*it, m_trucks[a]->boundingBox, scale))
 				return true;
 	} else if (m_trucks[b]->collisionBoundingBoxes.empty())
 	{
 		for (std::vector<AxisAlignedBox>::iterator it = m_trucks[a]->collisionBoundingBoxes.begin(); it != m_trucks[a]->collisionBoundingBoxes.end(); ++it)
-			if (it->intersects(m_trucks[b]->boundingBox))
+			if (intersectionAABB(*it, m_trucks[b]->boundingBox, scale))
 				return true;
 	} else
 	{
 		for (std::vector<AxisAlignedBox>::iterator it_a = m_trucks[a]->collisionBoundingBoxes.begin(); it_a != m_trucks[a]->collisionBoundingBoxes.end(); ++it_a)
 			for (std::vector<AxisAlignedBox>::iterator it_b = m_trucks[b]->collisionBoundingBoxes.begin(); it_b != m_trucks[b]->collisionBoundingBoxes.end(); ++it_b)
-				if (it_a->intersects(*it_b))
+				if (intersectionAABB(*it_a, *it_b, scale))
 					return true;
 	}
 
 	return false;
 }
 
-bool BeamFactory::predictTruckIntersectionCollAABB(int a, int b)
+bool BeamFactory::predictTruckIntersectionCollAABB(int a, int b, float scale)
 {
 	if (m_trucks[a]->predictedCollisionBoundingBoxes.empty() && m_trucks[b]->predictedCollisionBoundingBoxes.empty())
 	{
-		return predictTruckIntersectionAABB(a, b);
+		return predictTruckIntersectionAABB(a, b, scale);
 	} else if (m_trucks[a]->predictedCollisionBoundingBoxes.empty())
 	{
 		for (std::vector<AxisAlignedBox>::iterator it = m_trucks[b]->predictedCollisionBoundingBoxes.begin(); it != m_trucks[b]->predictedCollisionBoundingBoxes.end(); ++it)
-			if (it->intersects(m_trucks[a]->predictedBoundingBox))
+			if (intersectionAABB(*it, m_trucks[a]->predictedBoundingBox, scale))
 				return true;
 	} else if (m_trucks[b]->predictedCollisionBoundingBoxes.empty())
 	{
 		for (std::vector<AxisAlignedBox>::iterator it = m_trucks[a]->predictedCollisionBoundingBoxes.begin(); it != m_trucks[a]->predictedCollisionBoundingBoxes.end(); ++it)
-			if (it->intersects(m_trucks[b]->predictedBoundingBox))
+			if (intersectionAABB(*it, m_trucks[b]->predictedBoundingBox, scale))
 				return true;
 	} else
 	{
 		for (std::vector<AxisAlignedBox>::iterator it_a = m_trucks[a]->predictedCollisionBoundingBoxes.begin(); it_a != m_trucks[a]->predictedCollisionBoundingBoxes.end(); ++it_a)
 			for (std::vector<AxisAlignedBox>::iterator it_b = m_trucks[b]->predictedCollisionBoundingBoxes.begin(); it_b != m_trucks[b]->predictedCollisionBoundingBoxes.end(); ++it_b)
-				if (it_a->intersects(*it_b))
+				if (intersectionAABB(*it_a, *it_b, scale))
 					return true;
 	}
 
 	return false;
 }
 
-void BeamFactory::RecursiveActivation(int j)
+void BeamFactory::RecursiveActivation(int j, std::bitset<MAX_TRUCKS> &visited)
 {
 	if (!m_trucks[j] || m_trucks[j]->state != SIMULATED) return;
 
+	visited.set(j, true);
+
 	for (int t=0; t < m_free_truck; t++)
 	{
-		if (t == j || !m_trucks[t]) continue;
-		if (predictTruckIntersectionCollAABB(t, j))
+		if (t == j || !m_trucks[t] || visited[t]) continue;

```

Comment:
Doesn't this completely defeat the point of the recursion? Unless I'm missing something, this means each truck will be evaluated exactly 1x no matter how many trucks may actually intersect with it, and that's a broken design.